src/main/java/org/elasticsearch/search/aggregations/bucket/DeferringBucketCollector.java

```diff
@@ -138,6 +138,9 @@ public void replay(long... selectedBuckets) throws IOException {
         this.selectedBuckets = hash;
 
         collector.preCollection();
+        if (collector.needsScores()) {
+            throw new ElasticsearchIllegalStateException("Cannot defer if scores are needed");
+        }
```

Comment:
Is there any way to do this check earlier? maybe in the preCollection method? Only wondering as we will have already done the main collect phase by this point so it would be nice to fail faster if we can